include/RAJA/pattern/atomic.hpp

```diff
@@ -307,29 +307,29 @@ class AtomicRef
 
   RAJA_INLINE
   RAJA_HOST_DEVICE
-  void store(value_type rhs) const
+  void store(value_type rhs) const volatile
   {
     *m_value_ptr = rhs;
   }
 
   RAJA_INLINE
   RAJA_HOST_DEVICE
-  value_type operator=(value_type rhs) const
+  value_type operator=(value_type rhs) const volatile
   {
     *m_value_ptr = rhs;
     return rhs;
   }
 
   RAJA_INLINE
   RAJA_HOST_DEVICE
-  value_type load() const
+  value_type load() const volatile
   {
     return *m_value_ptr;
   }
 
   RAJA_INLINE
   RAJA_HOST_DEVICE
-  operator value_type() const
+  operator value_type() const volatile
```

Comment:
Should these methods be volatile.

Reply:
I changed `m_value_pointer` to be non-volatile since the atomic methods now accept a non-volatile pointer. The volatile qualification is recovered once you get to the specific policy implementation (except the serial implementation). However when `AtomicRef` access `m_value_pointer` directly I believe it should be volatile which is accomplished through the method qualification.

Reply:
I see what you mean, however adding volatile to the method will make the pointer instead of what it points to volatile.
Just removing volatile may be the way to go here.

Reply:
Duh. Would that be preferable to changing `m_value_ptr` to a reference in which case the `volatile` would propagate correctly?

Reply:
@MrBurmark let me know which you'd like to go and I'll make the fix.

Reply:
My current thinking is that volatile should be removed because atomic is not the same thing as volatile. The intention of this class is to be non-atomic in the seq case, so it should be subject to seq optimization in that case.
So my recommendation is to add policied atomic load and store functions and use them here instead of volatile.

Reply:
Sounds good to me.

Reply:
I removed the volatile qualifier from the methods but I didn't add policy specializations for the load and store. After reading a bit about volatile memory it's not clear to me that it does anything useful when implementing the atomics. However I'd be happy to add in policy specializations.

Reply:
This looks good. It makes sense for each function in atomicRef to be implemented via a policy specialized function.

Reply:
Done.

Reply:
This looks good, but I'm concerned that the atomic load and store don't actually use atomics to perform their function. I've used atomic exchange for store and atomicCAS with the same value for load before. Does that seem reasonable to you @trws.